base/nullable.jl

```diff
@@ -217,3 +217,127 @@ function hash(x::Nullable, h::UInt)
         return hash(x.value, h + nullablehash_seed)
     end
 end
+
+"""
+    Lifted{F}
+
+A type used to represent the lifted version of a function `f::F`.
+
+Calling an `_f::Lifted{F}` on arguments `xs...` lowers to
+`lift(_f.f, U, xs...)`, where the return type parameter `U` is chosen with the
+help of type inference.
+"""
+immutable Lifted{F}
+    f::F
+    cache::Dict{Tuple{Vararg{DataType}}, DataType}
+
+    (::Type{Lifted}){F}(f::F) = new{F}(
+        f, Dict{Tuple{Vararg{DataType}}, DataType}()
+    )
+end
+
+function (_f::Lifted{F}){F}(xs...)
+    f, cache = _f.f, _f.cache
+    signature = map(eltype, xs)
+    U = Base.@get!(
+        cache,
+        signature,
+        Core.Inference.return_type(f, Tuple{signature...})
+    )
+    return lift(f, U, xs...)
+end
+
+"""
+    lift(f::F)::Lifted{F}
+
+Return a lifted version of `f`.
+"""
+lift(f) = Lifted(f)
+
+"""
+    lift(f, U, xs...)
+
+Return an empty `Nullable{U}` if any of the `xs` is null; otherwise, return the
+(`Nullable`-wrapped) value of `f` applied to the values of the `xs`.
+
+NOTE: There are two exceptions to the above: `lift(|, Bool, x, y)` and
+`lift(&, Bool, x, y)`. These methods both follow three-valued logic semantics.
+"""
+function lift(f, U::DataType, x)
+    if isnull(x)
+        return Nullable{U}()
+    else
+        return Nullable{U}(f(unsafe_get(x)))
+    end
+end
+
+function lift(f, U::DataType, x1, x2)
+    if isnull(x1) | isnull(x2)
+        return Nullable{U}()
+    else
+        return Nullable{U}(f(unsafe_get(x1), unsafe_get(x2)))
+    end
+end
+
+function lift(f, U::DataType, xs...)
+    if mapreduce(isnull, |, false, xs)
+        return Nullable{U}()
+    else
+        return Nullable{U}(f(map(unsafe_get, xs)...))
+    end
+end
+
+# Three-valued logic
+
+(::Lifted{&})(x::Union{Bool, Nullable{Bool}}, y::Union{Bool, Nullable{Bool}}) =
+    lift(&, Bool, x, y)
+(::Lifted{|})(x::Union{Bool, Nullable{Bool}}, y::Union{Bool, Nullable{Bool}}) =
+    lift(|, Bool, x, y)
+
+function lift(f::typeof(&), ::Type{Bool}, x, y)::Nullable{Bool}
+    return ifelse(
+        isnull(x),
+        ifelse(
+            isnull(y),
+            Nullable{Bool}(),
```

Comment:
this uses a lot of vertical space, would be better with more than a single token per line